In [211]:
import pandas as pd
import numpy as np
import random
import sklearn
import sklearn.model_selection
import sklearn.linear_model
import sklearn.metrics
import sklearn.metrics.pairwise
import math

ratings = pd.read_csv('./ml-latest-small/ratings.csv')
movies = pd.read_csv('./ml-latest-small/movies.csv')
tags = pd.read_csv('./ml-latest-small/tags.csv')

#Split 80%/20% between train and test
train_ratings, test_ratings = sklearn.model_selection.train_test_split(ratings, test_size=0.2)

Content-based Filtering
============
Content-based filtering methods are based on a description of the item and a profile of the user’s preferences

Uses user and item features to create a recommendation model. Think of it as like any regression or classification problem in machine learning

Example: Linear Regression
------

We'll use the movie genres as the movie features. Process it as one hot encoding vector for each movie

In [46]:
#Process movie features
genres = set()
movie_genres = {}
for index, row in movies.iterrows():
    genres.add(row['genres'])
    movie_genres[row['movieId']] = row['genres'].split("|")
    
genres_a = []
for g in genres:
    genres_a.append(g)
    
movie_features = {}
for key in movie_genres:
    val = movie_genres[key]
    movie_features[key] = []
    for g in genres_a:
        movie_features[key].append(1 if g in val else 0)

In [58]:
lr = sklearn.linear_model.LinearRegression()

train_features = []
train_labels = []

test_features = []
test_labels = []

for index, row in train_ratings.iterrows():
    train_features.append(movie_features[row['movieId']])
    train_labels.append(row['rating'])
    
for index, row in test_ratings.iterrows():
    test_features.append(movie_features[row['movieId']])
    test_labels.append(row['rating'])
    
lr.fit(train_features, train_labels)

LinearRegression(copy_X=True, fit_intercept=True, n_jobs=1, normalize=False)

Next step is to evaluate our models. A common metric for evaluating recommendation algorithms is the Root Mean Squared Error (RMSE). 

$$RMSE = \sqrt{\frac{1}{N} \sum\limits_{i}^{N} (y_i-\hat{y}_i)^2}$$

In [67]:
predicted_ratings = lr.predict(test_features)

rmse = math.sqrt(sklearn.metrics.mean_squared_error(predicted_ratings, test_labels))
print "RMSE:", rmse

RMSE: 1.03700622971


An un-regularized linear regression model with just using movie genres is getting an RMSE 1.037! That's not too bad already. You can probably even continue to try to improve this model by incorporating some user features (not available in movielens 100k) or adding regularization to the regression

Collaborative Filtering
============

In collaborative filtering, we can make recommendations using only information about the past preferences of the the user along with other users. We don't need any user or item features, and it generally works better than content-based recommendation.

Visualization:
------
<img src="./images/Collaborative_filtering-14.tiff" style="height:300px;width:300px"/>

<img src="./images/Collaborative_filtering-16.tiff" style="height:300px;width:300px"/>

K-Nearest Neighbors
--------------------

A common form of collaborative filtering is the nearest neighbors method. The intuition behind it is:
1. Get the similarities between users or items (More on this later)
2. Weigh the recommendations of the nearest neighbors of the user based on the similarities

In [101]:
#Get lookup dict or ratings from movie to user perspective
movie_users = {}
user_movies = {}
for index, row in train_ratings.iterrows():
    if row['movieId'] not in movie_users:
        movie_users[row['movieId']] = {}
        
    if row['userId'] not in user_movies:
        user_movies[row['userId']] = {}
        
    movie_users[row['movieId']][row['userId']] = row['rating']
    user_movies[row['userId']][row['movieId']] = row['rating']
     

We need a similarity metric. A simple one we can use is the cosine similarity between two vectors A and B. In this setting A and B are the ratings each user has given to movies that they have <b>both watched</b>.

$$\cos(A,B)=\frac{\sum\limits_{i}^{N}A_iB_i}{\sqrt{\sum\limits_{i}^{N}A_i^2} \sqrt{\sum\limits_{i}^{N}B_i^2}} $$

In [200]:
#Calculate similarity ratings for all user combinations
#Uses cosine similarity
similarities = {}

for user1 in users:
    if user1 not in similarities:
        similarities[user1] = {}
    
    for user2 in users:
        if user2 in similarities[user1] or user1 == user2:
            continue
        
        user1_ratings = []
        user2_ratings = []
        
        for key in user_movies[user1]:
            watched_users = movie_users[key]
            if user1 in watched_users and user2 in watched_users:
                user1_ratings.append(watched_users[user1])
                user2_ratings.append(watched_users[user2])
                
        if len(user1_ratings) > 0:
            if user2 not in similarities:
                similarities[user2] = {}
            
            sim = np.dot(user1_ratings, user2_ratings) / (np.linalg.norm(user1_ratings) * np.linalg.norm(user2_ratings))
            similarities[user1][user2] = sim
            similarities[user2][user1] = sim
            

Using the similarities, we can use nearest neighbor regression using the following formula: 

$$R_{x,y} = \frac{\sum\limits_{z} Sim_{x,z}R_{z,y}}{\sum\limits_{z} Sim_{x,z}}$$

In [172]:
#Nearest neighbor regression
def knn_predict(k, userId, movieId):
    watched_users = movie_users[movieId]
    
    sim_scores = []
    rating_scores = []
    
    for w in watched_users:
        if w in similarities[userId]:
            sim_scores.append(similarities[userId][w])
            rating_scores.append(watched_users[w])
    
    #Got lazy. Used pandas DataFrames to sort by similarity and get the nearest neighbors
    if len(sim_scores) > 0:
        sim_df = pd.DataFrame(data={'sim' : sim_scores, 'rating' : rating_scores})
        sorted_sim_df = sim_df.sort_values(by=['sim'], ascending = False)
        knearest = sorted_sim_df.head(k)
        
        numerator = 0
        denomenator = 0
        for index, row in knearest.iterrows():
            #print row['sim'], row['rating']
            numerator += row['sim'] * row['rating']
            denomenator += row['sim']
        #print "Quotient", numerator, denomenator, numerator / denomenator  
        return numerator / denomenator
    else: #What to do if user has no similarities??
        return 2.5
    
    


In [180]:
predicted_ratings = []
actual_ratings = []
K = 5
for index, row in test_ratings.iterrows():
    if row['movieId'] not in movie_users:
        continue
        
    pred = knn_predict(50, row['userId'], row['movieId'])
    predicted_ratings.append(pred)
    actual_ratings.append(row['rating'])
    
rmse = math.sqrt(sklearn.metrics.mean_squared_error(predicted_ratings, actual_ratings))
print "RMSE:", rmse

No sim
No sim
RMSE: 0.985502162215


RMSE of 0.98! Already better than the linear regression example, but hopefully can still be improved. Two main things you can fiddle with here aside from K:

1. Use item similarity instead of user. This was pioneered by Amazon in 1998 and documented in papers they published in 2001 and 2003 (https://www.cs.umd.edu/~samir/498/Amazon-Recommendations.pdf). This works especially better when there are fewer items than users, so similarities will be richer due to more intersections.

2. Try different similarity metrics. For example: Pearson correlation which goes from -1 to 1 instead of just 0 to 1.

Matrix Factorization
--------

A more powerful form of collaborative filtering is matrix factorization. The basic intuition is that it models the user-item ratings/preferences as an $M \times N$ matrix $R$ where $M$ is the number of users and $N$ is the number of items.
<table>
    <tr>
        <th></th>
        <th>Movie1</th>
        <th>Movie2</th>
        <th>Movie3</th>
        <th>Movie4</th>
    </tr>
    <tr>
        <th>User1</th>
        <td>5</td>
        <td></td>
        <td>1.5</td>
        <td>3.5</td>
    </tr>
    <tr>
        <th>User2</th>
        <td>1</td>
        <td>3.5</td>
        <td></td>
        <td>4</td>
    </tr>
    <tr>
        <th>User3</th>
        <td></td>
        <td>2</td>
        <td>4</td>
        <td>5</td>
    </tr>
</table>

There are different kinds of matrix factorization methods for CF. A simple one is Probabilistic Matrix Factorization (https://papers.nips.cc/paper/3208-probabilistic-matrix-factorization.pdf). In PMF the ratings matrix $R$ is decomposed into an $M \times K$ matrix $U$ and a $N \times K$ matrix $V$ such that

$$ R = U^TV$$

### Learning $U$ and $V$ ######

So how do we learn U and V? We can do gradient descent using Alternating Least Squares to minimize the following error function:

$$ 
E = \frac{1}{2} \sum\limits_{i}^{N}\sum\limits_{j}^{M}(R_{i,j} - U_{i}^{T}V_{j})^2 
+ \frac{\lambda_U}{2} \sum\limits_{i}^{N} \|{U_i}\|_{Fro}^2
+ \frac{\lambda_V}{2} \sum\limits_{j}^{M} \|{V_j}\|_{Fro}^2
$$

In [ ]:
def get_error(user_matrix, item_matrix, actual, lu, li):
    error = 0
    for u in actual:
        for i in actual[u]:
            pred = np.dot(user_matrix[u], item_matrix[i])
            error += (pred - actual[u][i]) ** 2
            
    error /= 2
    
    u_norm = 0
    for u in user_matrix:
        for val in user_matrix[u]:
            u_norm += val ** 2
    i_norm = 0
    for i in item_matrix:
        for val in item_matrix[i]:
            i_norm += val ** 2
            
    error += lu * u_norm / 2
    error += li * i_norm / 2
              
    return error
    

In [141]:
#Derivative with respect to U
def dE_u(user_matrix, item_matrix, actual, lu, u, index):
    derivative = 0
    for i in actual[u]:
        pred = np.dot(user_matrix[u], item_matrix[i])
        derivative += (pred - actual[u][i]) * item_matrix[i][index]
        
    derivative += lu * user_matrix[u][index]
    
    return derivative

In [146]:
#Derivative with respect to V
def dE_i(user_matrix, item_matrix, actual, li, i, index):
    derivative = 0
    watched_users = movie_users[i]
    for u in watched_users:
        pred = np.dot(user_matrix[u], item_matrix[i])
        derivative += (pred - actual[u][i]) * user_matrix[u][index]
        
    derivative += li * item_matrix[i][index]
    
    return derivative

In Alternating Least Squares, we get the gradients of $E$ by fixing $V$ as constants and getting the derivative of $E$ with respect to $U$ and vice versa. We alternate between both gradients when we update the matrices.

In [192]:
#Alternating Least Squares
def learn(user_matrix, item_matrix, actual, lu, li, learning_rate):
    counter = 0
    while counter < 1000:
        error = get_error(user_matrix, item_matrix, actual, lu, li)
        print "Counter:", counter, "Error:", error #Error should trend downwards over time
        
        #Treat V as contant, update U with its gradients
        if counter % 2 == 0:
            user_derivatives = {}      
            for u in user_matrix:
                user_derivatives[u] = []
            
                for index in range(len(user_matrix[u])):
                    user_derivatives[u].append(dE_u(user_matrix, item_matrix, actual, lu, u, index))
                    
            for u in user_matrix:
                for index in range(len(user_matrix[u])):
                    user_matrix[u][index] -= learning_rate * user_derivatives[u][index]

        #Treat U as constant, update V with its gradients
        else:
            item_derivatives = {}
            for i in item_matrix:
                item_derivatives[i] = []
                for index in range(len(item_matrix[i])):
                    item_derivatives[i].append(dE_i(user_matrix, item_matrix, actual, li, i, index))
                    
            for i in item_matrix:
                for index in range(len(item_matrix[i])):
                    item_matrix[i][index] -= learning_rate * item_derivatives[i][index]
        
        counter += 1
        
    return (user_matrix, item_matrix)

In [217]:
user_matrix = {}
item_matrix = {}
learning_rate = 0.0001
lu = 0.0001
li = 0.0001
K = 5

for u in user_movies:
    user_matrix[u] = random.sample(range(1, 100), K)
    for index in range(K):
        user_matrix[u][index] /= float(100)
    
for i in movie_users:
    item_matrix[i] = random.sample(range(1, 100), K)
    for index in range(K):
        item_matrix[i][index] /= float(100)
    
learned_users, learned_items = learn(user_matrix, item_matrix, user_movies, lu, li, learning_rate)

Counter: 0 Error: 266592.87871137116
Counter: 1 Error: 249180.87811246203
Counter: 2 Error: 245861.01961888152
Counter: 3 Error: 231680.55906254693
Counter: 4 Error: 228340.52320349592
Counter: 5 Error: 216498.91192437397
Counter: 6 Error: 213193.31768429573
Counter: 7 Error: 203103.8043315066
Counter: 8 Error: 199874.44764229993
Counter: 9 Error: 191141.26278308933
Counter: 10 Error: 188019.51925077647
Counter: 11 Error: 180367.98512662304
Counter: 12 Error: 177376.66112589254
Counter: 13 Error: 170610.21416396773
Counter: 14 Error: 167765.00668818367
Counter: 15 Error: 161738.46574783526
Counter: 16 Error: 159049.163692247
Counter: 17 Error: 153651.89910242145
Counter: 18 Error: 151123.41181419574
Counter: 19 Error: 146268.60605881584
Counter: 20 Error: 143901.8676089977
Counter: 21 Error: 139519.56170614136
Counter: 22 Error: 137312.33503466926
Counter: 23 Error: 133344.85261881107
Counter: 24 Error: 131292.4439129323
Counter: 25 Error: 127691.32531499949
Counter: 26 Error: 125787.2

Counter: 217 Error: 41262.9756223672
Counter: 218 Error: 41192.65318694694
Counter: 219 Error: 41118.018467681395
Counter: 220 Error: 41048.78103794306
Counter: 221 Error: 40975.53608606888
Counter: 222 Error: 40907.35730593184
Counter: 223 Error: 40835.46641426696
Counter: 224 Error: 40768.32077173021
Counter: 225 Error: 40697.749426696384
Counter: 226 Error: 40631.61222069322
Counter: 227 Error: 40562.327052384266
Counter: 228 Error: 40497.174361161364
Counter: 229 Error: 40429.14309594695
Counter: 230 Error: 40364.95174689765
Counter: 231 Error: 40298.14316240749
Counter: 232 Error: 40234.89070327971
Counter: 233 Error: 40169.27458562821
Counter: 234 Error: 40106.93925703899
Counter: 235 Error: 40042.486360153875
Counter: 236 Error: 39981.047069342174
Counter: 237 Error: 39917.72907628453
Counter: 238 Error: 39857.16537202612
Counter: 239 Error: 39794.95485817657
Counter: 240 Error: 39735.24690681831
Counter: 241 Error: 39674.11730483418
Counter: 242 Error: 39615.245867386875
Counte

Counter: 432 Error: 33147.53919876345
Counter: 433 Error: 33131.02468467927
Counter: 434 Error: 33108.7065944419
Counter: 435 Error: 33092.37070528005
Counter: 436 Error: 33070.21633042704
Counter: 437 Error: 33054.05645915248
Counter: 438 Error: 33032.063743908286
Counter: 439 Error: 33016.07733107551
Counter: 440 Error: 32994.244255238926
Counter: 441 Error: 32978.42878793686
Counter: 442 Error: 32956.75336610907
Counter: 443 Error: 32941.10637694091
Counter: 444 Error: 32919.586657780026
Counter: 445 Error: 32904.105723862165
Counter: 446 Error: 32882.73978934861
Counter: 447 Error: 32867.422531344295
Counter: 448 Error: 32846.208496068204
Counter: 449 Error: 32831.05257723783
Counter: 450 Error: 32809.98858770748
Counter: 451 Error: 32794.99171298841
Counter: 452 Error: 32774.075946953046
Counter: 453 Error: 32759.235862058897
Counter: 454 Error: 32738.466527848945
Counter: 455 Error: 32723.781018395046
Counter: 456 Error: 32703.156354279305
Counter: 457 Error: 32688.623244929622
C

Counter: 645 Error: 30315.422886720287
Counter: 646 Error: 30303.38791827506
Counter: 647 Error: 30297.139380675857
Counter: 648 Error: 30285.16024627458
Counter: 649 Error: 30278.95891015032
Counter: 650 Error: 30267.03515354033
Counter: 651 Error: 30260.88051924051
Counter: 652 Error: 30249.01168940279
Counter: 653 Error: 30242.903263766417
Counter: 654 Error: 30231.088914835935
Counter: 655 Error: 30225.026211083325
Counter: 656 Error: 30213.265902273728
Counter: 657 Error: 30207.24843991183
Counter: 658 Error: 30195.541735440995
Counter: 659 Error: 30189.569040163104
Counter: 660 Error: 30177.91550917791
Counter: 661 Error: 30171.987112770214
Counter: 662 Error: 30160.386329273948
Counter: 663 Error: 30154.50176952119
Counter: 664 Error: 30142.953312304347
Counter: 665 Error: 30137.112132896917
Counter: 666 Error: 30125.615585464722
Counter: 667 Error: 30119.81733590735
Counter: 668 Error: 30108.372286414124
Counter: 669 Error: 30102.61652193947
Counter: 670 Error: 30091.2225631178

Counter: 858 Error: 28802.755301210276
Counter: 859 Error: 28799.61637599281
Counter: 860 Error: 28791.714467531296
Counter: 861 Error: 28788.592747320537
Counter: 862 Error: 28780.717090066155
Counter: 863 Error: 28777.612427760614
Counter: 864 Error: 28769.7628612044
Counter: 865 Error: 28766.675111201454
Counter: 866 Error: 28758.851476157583
Counter: 867 Error: 28755.78049433057
Counter: 868 Error: 28747.982632922034
Counter: 869 Error: 28744.928276605806
Counter: 870 Error: 28737.156032249884
Counter: 871 Error: 28734.118160221347
Counter: 872 Error: 28726.371377613745
Counter: 873 Error: 28723.349850077488
Counter: 874 Error: 28715.62837518159
Counter: 875 Error: 28712.623053746916
Counter: 876 Error: 28704.926733779488
Counter: 877 Error: 28701.937481449786
Counter: 878 Error: 28694.266164864002
Counter: 879 Error: 28691.292846016546
Counter: 880 Error: 28683.64638249268
Counter: 881 Error: 28680.688862861985
Counter: 882 Error: 28673.067103294576
Counter: 883 Error: 28670.12524

In [219]:
predicted_ratings = []
actuals = []
not_count = 0
for index, row in test_ratings.iterrows():
    if row['movieId'] not in movie_users:
        continue
        
    pred = np.dot(learned_users[row['userId']], learned_items[row['movieId']])
    predicted_ratings.append(pred)
    actuals.append(row['rating'])

rmse = math.sqrt(sklearn.metrics.mean_squared_error(predicted_ratings, actuals))
print "RMSE:", rmse

RMSE: 0.858861345016


<table>
    <tr>
        <th colspan=2><center>K=5</center></th>
    <tr>
        <th>$\lambda$</th>
        <th>RMSE</th>
    </tr>
    <tr>
        <td>0.0001</td>
        <td>0.858861345016</td>
    <tr>
        <td>0.001</td>
        <td>0.913146934761</td>
    </tr>
    <tr>
        <td>0.01</td>
        <td>0.912297969849</td>
    </tr>
    <tr>
        <td>0.1</td> 
        <td>0.913277052092</td>
    </tr>
    <tr>
        <td>1</td>
        <td>0.916509506566</td>
    </tr>
    <tr>
        <td>10</td>
        <td>1.01069785051</td>
    </tr>
</table>

<table>
    <tr>
        <th colspan=2><center>K=10</center></th>
    <tr>
        <th>$\lambda$</th>
        <th>RMSE</th>
    </tr>
    <tr>
        <td>0.0001</td>
        <td>0.91227081992</td>
    </tr>
    <tr>
        <td>0.001</td>
        <td>0.908526077502</td>
    </tr>
    <tr>
        <td>0.01</td>
        <td>0.910726843867</td>
    </tr>
    <tr>
        <td>0.1</td> 
        <td>0.846665725031</td>
    </tr>
    <tr>
        <td>1</td>
        <td>0.848858193242</td>
    </tr>
    <tr>
        <td>10</td>
        <td>0.927015692586</td>
    </tr>
</table>

Next Steps
==========

1. Various methods have been devised to combine collaborative filtering and content-based recommendations into hybrid recommender systems (For example: https://www.microsoft.com/en-us/research/wp-content/uploads/2009/01/www09.pdf). 

2. Session-based click stream can be used instead of past user preferences, using RNNs (https://arxiv.org/pdf/1511.06939.pdf).

3. Long-tail recommendations are hard. Recommendation engines tend to recommend the most popular items. 

4. Use frameworks :) Faster and easier. http://surpriselib.com